### PR TITLE
[IMPROVE] Replace the Department select component by an Autocomplete input in Omnichannel UI

### PR DESCRIPTION
**Pull request by renatobecker**
- Added a new endpoint to fetch the data through the Autocomplete component
- Replaced the Department Select component with an Autocomplete input in the Current Chat window
- Replaced the Department Select component with an Autocomplete input in the Forward Chat panel

How does this look like now:

**Current Chats list:**
Screen Shot 2020-02-20 at 15 58 35

**Forward Chat panel:**

Screen Shot 2020-02-20 at 15 59 08
Screen Shot 2020-02-20 at 15 59 21

### CODE REVIEW

File: app/livechat/client/views/app/tabbar/visitorForward.js [hunk 1]
```diff
@@ -43,13 +40,41 @@ Template.visitorForward.helpers({
 	onClickTagAgent() {
 		return Template.instance().onClickTagAgent;
 	},
+	departmentModifier() {
+		return (filter, text = '') => {
+			const f = filter.get();
+			return `${ f.length === 0 ? text : text.replace(new RegExp(filter.get()), (part) => `<strong>${ part }</strong>`) }`;
```
Comment (rodrigok) [inline app/livechat/client/views/app/tabbar/visitorForward.js]:
```suggestion
			return `${ f.length === 0 ? text : text.replace(new RegExp(filter.get(), 'i'), (part) => `<strong>${ part }</strong>`) }`;
```

File: app/livechat/client/views/app/livechatCurrentChats.js [hunk 2]
```diff
@@ -84,12 +84,24 @@ Template.livechatCurrentChats.helpers({
...
-	departments() {
-		return Template.instance().departments.get();
-	},
 	tagId() {
 		return this;
 	},
+	departmentModifier() {
+		return (filter, text = '') => {
+			const f = filter.get();
+			return `${ f.length === 0 ? text : text.replace(new RegExp(filter.get()), (part) => `<strong>${ part }</strong>`) }`;
```
Comment (rodrigok) [inline app/livechat/client/views/app/livechatCurrentChats.js]:
```suggestion
			return `${ f.length === 0 ? text : text.replace(new RegExp(filter.get(), 'i'), (part) => `<strong>${ part }</strong>`) }`;
```

File: app/livechat/client/views/app/tabbar/visitorForward.js [hunk 3]
```diff
@@ -83,8 +108,9 @@ Template.visitorForward.events({
 		const [user] = instance.selectedAgents.get();
 		if (user) {
 			transferData.userId = user._id;
-		} else if (instance.find('#forwardDepartment').value) {
-			transferData.departmentId = instance.find('#forwardDepartment').value;
+		} else if (instance.selectedDepartments.get()) {
+			const [department] = instance.selectedDepartments.get();
+			transferData.departmentId = department._id;
```
Comment (rodrigok) [inline app/livechat/client/views/app/tabbar/visitorForward.js]:
```suggestion
			transferData.departmentId = department && department._id;
```

Comment (rodrigok) [reply app/livechat/client/views/app/tabbar/visitorForward.js]: Should check if the departmentId or the userId was set before call the method below

File: app/livechat/client/views/app/tabbar/visitorForward.js [hunk 4]
```diff
@@ -101,14 +127,15 @@ Template.visitorForward.events({
 	},
 
 	'change #forwardDepartment, blur #forwardDepartment'(event, instance) {
-		if (event.currentTarget.value) {
-			instance.find('#forwardUser').value = '';
+		const [user] = instance.selectedAgents.get();
+		if (user) {
+			instance.selectedAgents.set([]);
```
Comment (rodrigok) [inline app/livechat/client/views/app/tabbar/visitorForward.js]: This is not working

File: app/livechat/client/views/app/tabbar/visitorForward.js [hunk 5]
```diff
@@ -101,14 +127,15 @@ Template.visitorForward.events({
...
 		}
 	},
 
 	'change #forwardUser, blur #forwardUser'(event, instance) {
 		if (event.currentTarget.value && instance.find('#forwardDepartment')) {
-			instance.find('#forwardDepartment').value = '';
+			instance.selectedDepartments.set([]);
```
Comment (rodrigok) [inline app/livechat/client/views/app/tabbar/visitorForward.js]: This is not working

File: app/livechat/server/api/lib/departments.js [hunk 6]
```diff
@@ -44,3 +44,26 @@ export async function findDepartmentById({ userId, departmentId, includeAgents =
 
 	return result;
 }
+
+export async function findDepartmentsToAutocomplete({ uid, selector }) {
+	if (!await hasPermissionAsync(uid, 'view-livechat-departments') && !await hasPermissionAsync(uid, 'view-l-room')) {
+		return { items: [] };
+	}
+	const exceptions = selector.exceptions || [];
+	const conditions = selector.conditions || {};
```
Comment (rodrigok) [inline app/livechat/server/api/lib/departments.js]:
```suggestion
	const { exceptions = [], conditions = {} } = selector;
```

File: app/models/server/raw/LivechatDepartment.js [hunk 7]
```diff
@@ -1,8 +1,32 @@
...
 export class LivechatDepartmentRaw extends BaseRaw {
 	findInIds(departmentsIds, options) {
 		const query = { _id: { $in: departmentsIds } };
 		return this.find(query, options);
 	}
+
+	findByNameRegexWithExceptionsAndConditions(searchTerm, exceptions, conditions, options) {
+		if (exceptions == null) { exceptions = []; }
+		if (conditions == null) { conditions = {}; }
+		if (options == null) { options = {}; }
```
Comment (rodrigok) [inline app/models/server/raw/LivechatDepartment.js]:
```suggestion
	findByNameRegexWithExceptionsAndConditions(searchTerm, exceptions = [], conditions = {}, options = {}) {
```

File: app/models/server/raw/LivechatDepartment.js [hunk 8]
```diff
@@ -1,8 +1,32 @@
...
+		if (!_.isArray(exceptions)) {
```
Comment (rodrigok) [inline app/models/server/raw/LivechatDepartment.js]:
```suggestion
		if (!Array.isArray(exceptions)) {
```

File: app/models/server/raw/LivechatDepartment.js [hunk 9]
```diff
@@ -1,8 +1,32 @@
+import _ from 'underscore';
```
Comment (rodrigok) [inline app/models/server/raw/LivechatDepartment.js]:
```suggestion
```

File: app/livechat/client/views/app/tabbar/visitorForward.js (hunk 3 above)
Comment (renatobecker) [reply app/livechat/client/views/app/tabbar/visitorForward.js]: Done.

File: app/livechat/client/views/app/tabbar/visitorForward.js (hunk 4 above)
Comment (renatobecker) [reply app/livechat/client/views/app/tabbar/visitorForward.js]: Done. It was implemented to listen to the old select component.

File: app/livechat/client/views/app/tabbar/visitorForward.js (hunk 5 above)
Comment (renatobecker) [reply app/livechat/client/views/app/tabbar/visitorForward.js]: Done. It was implemented to listen to the old select component.

**Pull request merged**